File: array/1002.py
class Solution:
    def commonChars(self, A: List[str]) -> List[str]:
        words_list = []
        for word in A:
            word_list = list(word)
            words_list.append(word_list)
            
        ret = []
        for c in words_list[0]:
            flag = 0
            for i in range(1, len(words_list)):
                if not c in words_list[i]:
                    flag = 1
                    break
                else:
                    words_list[i].remove(c)
            if flag == 0:
                ret.append(c)
        
        return ret
            

File: hashtable/1002.java
import java.util.*;
class Solution {
    public List<String> commonChars(String[] A) {
        List<String> ret = new ArrayList<>();
        List<Map<String, Integer>> list = new ArrayList<>();
        for (String s : A) {
            Map<String, Integer> map = new HashMap<>();
            for (int i = 0; i < s.length(); i++) {
                String c = String.valueOf(s.charAt(i));
                if (map.containsKey(c)) {
                    map.put(c, map.get(c) + 1);
                } else {
                    map.put(c, 1);
                }
            }
            list.add(map);
        }
        Map<String, Integer> targetMap = list.get(0);
        for (Map.Entry<String, Integer> entry : targetMap.entrySet()) {
            String key = entry.getKey();
            int value = entry.getValue();
            int repeat = value;
            boolean flag = true;
            for (int i = 1; i < list.size(); i++) {
                if (list.get(i).containsKey(key)) {
                    repeat = Math.min(repeat, list.get(i).get(key));
                } else {
                    flag = false;
                    break;
                }
            }
            while (flag && repeat > 0) {
                ret.add(key);
                repeat--;
            }            
        }
        return ret;
    }
}
